src/datatablerow/utils.jl

```diff
@@ -0,0 +1,191 @@
+# Rows grouping.
+# Maps row contents to the indices of all the equal rows.
+# Used by groupby(), join(), nonunique()
+immutable RowGroupDict{T<:AbstractDataTable}
+    "source data table"
+    dt::T
+    "number of groups"
+    ngroups::Int
+    "row hashes"
+    rhashes::Vector{UInt}
+    "hashindex -> index of group-representative row"
+    gslots::Vector{Int}
+    "group index for each row"
+    groups::Vector{Int}
+    "permutation of row indices that sorts them by groups"
+    rperm::Vector{Int}
+    "starts of ranges in rperm for each group"
+    starts::Vector{Int}
+    "stops of ranges in rperm for each group"
+    stops::Vector{Int}
+end
+
+# "kernel" functions for hashrows()
+# adjust row hashes by the hashes of column elements
+function hashrows_col!(h::Vector{UInt}, v::AbstractVector)
+    @inbounds for i in eachindex(h)
+        h[i] = hash(v[i], h[i])
+    end
+    h
+end
+
+function hashrows_col!{T<:Nullable}(h::Vector{UInt}, v::AbstractVector{T})
+    @inbounds for i in eachindex(h)
+        h[i] = isnull(v[i]) ?
+               hash(Base.nullablehash_seed, h[i]) :
```

Comment:
I got hash errors after making these changes (and others probably downstream of this)
```julia
Test Failed
  Expression: dt_rowhashes == [hash(dr) for dr = eachrow(dt)]
   Evaluated: UInt64[0x1bdefb2976bd94c3,0xdd9fa2a42135ac50,0x1e6098864a87ed7d,0x1bdefb2976bd94c3,0xdd9fa2a42135ac50,0x0e42569badde05fc] == UInt64[0x1bdefb2976bd94c3,0xbb2ac60630b5eb56,0x1e6098864a87ed7d,0x1bdefb2976bd94c3,0xbb2ac60630b5eb56,0xd4ac22e6c0fc8065]
```
I think you're referencing [this line](https://github.com/JuliaLang/julia/blob/master/base/nullable.jl#L248), but `hash(Base.nullablehash_seed, h[i])` doesn't call that function since the first argument is not a `Nullable` but instead `Base.nullablehash_seed`.